pkg/reconciler/controller/jetstream/controller.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jetstream
+
+import (
+	"context"
+
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	deploymentinformer "knative.dev/pkg/client/injection/kube/informers/apps/v1/deployment"
+	"knative.dev/pkg/client/injection/kube/informers/core/v1/endpoints"
+	"knative.dev/pkg/client/injection/kube/informers/core/v1/service"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/system"
+
+	"k8s.io/client-go/tools/cache"
+
+	"knative.dev/eventing-natss/pkg/client/injection/informers/messaging/v1alpha1/natsjetstreamchannel"
+	jetstreamchannelreconciler "knative.dev/eventing-natss/pkg/client/injection/reconciler/messaging/v1alpha1/natsjetstreamchannel"
+)
+
+// NewController initializes the controller and is called by the generated code.
+// Registers event handlers to enqueue events.
+func NewController(ctx context.Context) *controller.Impl {
+
+	logger := logging.FromContext(ctx)
+	channelInformer := natsjetstreamchannel.Get(ctx)
+	deploymentInformer := deploymentinformer.Get(ctx)
+	serviceInformer := service.Get(ctx)
+	endpointsInformer := endpoints.Get(ctx)
+	kubeClient := kubeclient.Get(ctx)
+
+	r := &Reconciler{
+		kubeClientSet:            kubeClient,
+		dispatcherNamespace:      system.Namespace(),
+		dispatcherDeploymentName: dispatcherName,
+		dispatcherServiceName:    dispatcherName,
+		deploymentLister:         deploymentInformer.Lister(),
+		serviceLister:            serviceInformer.Lister(),
+		endpointsLister:          endpointsInformer.Lister(),
+	}
+
+	impl := jetstreamchannelreconciler.NewImpl(ctx, r)
+
+	logger.Info("Setting up event handlers")
+	channelInformer.Informer().AddEventHandler(controller.HandleAll(impl.Enqueue))
+
+	grCh := func(obj interface{}) {
+		impl.GlobalResync(channelInformer.Informer())
+	}
+	filterFunc := controller.FilterWithNameAndNamespace(r.dispatcherNamespace, r.dispatcherDeploymentName)
+
+	// Set up watches for dispatcher resources we care about, since any changes to these
+	// resources will affect our Channels. So, set up a watch here, that will cause
+	// a global Resync for all the channels to take stock of their health when these change.
+	deploymentInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: filterFunc,
+		Handler:    controller.HandleAll(grCh),
+	})
+	serviceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
```

Comment:
@vaikas we already have the eventhandler here https://github.com/knative-sandbox/eventing-natss/pull/182/files#diff-3b0c26560907627880663ead8f0ee94068572987d490b477626fa8e0868a7ff0R76

Reply:
Maybe I'm wrong. But this sets up the (what looks to be a single dispatcher for all the channels?) and that causes a global resync, which I think makes sense since if the dispatcher is used by all the channels, they all need to be reconciled. However, there's another Service thats created here (obvs if it doesn't exist):
https://github.com/knative-sandbox/eventing-natss/pull/182/files#diff-a9bd98b0e5248a4625de7f3a112db792e398a7fd49924ac9bd40bec7f0a788fdR20

For that I'd expect us to set up a tracker so that only one channel get's reconciled when that Service changes.

Reply:
@vaikas sorry for my above reply(a little Incorrect)。we now have two kind of service：
- `jetstream-ch-dispatcher` point to the deployment `jetstream-ch-dispatcher`. This one need global reconcile
- one service for one channel,such as `default-kne-trigger-kn-channel` which is `externalname` type.  Sorry for miss this kind, let me add eventhandler for this case.

Reply:
No worries :) I just as well could've missed something. Since you set up the OwnerRefs for it, I think you should just be able to add a IsControlledBy filter for Service informer for it. Something like this:
https://github.com/knative-sandbox/eventing-rabbitmq/blob/main/pkg/reconciler/trigger/controller.go#L117

But changing the types as appropriate. Key being using the Handler like so:

```
		Handler:    controller.HandleAll(impl.EnqueueControllerOf),
```
So that when that Service changes, it only enqueues the owning channel for reconcile, not a global resync.